templates/s3-bucket-v2.j2

```diff
@@ -169,37 +169,68 @@ Resources:
-            Effect: Deny
-            Principal: "*"
-            Action: s3:PutObject
-            Resource: !If [EnableEncryption, !Sub "${SynapseEncryptedExternalBucket.Arn}/*", !Sub "${SynapseExternalBucket.Arn}/*"]
-            Condition:
-              StringNotLike:
-                aws:userid:
-                  - arn:aws:iam::325565585839:root
-                  - !Ref AWS::AccountId
-              StringNotEquals:
-                s3:x-amz-acl: bucket-owner-full-control
-          -
-            Sid: "WriteAccess"
-            Effect: "Allow"
             Principal:
-              AWS: !Ref GrantAccess
+              AWS: "325565585839"
+          - Sid: SynapseObjectAccess
             Action:
               - !If [AllowWrite, "s3:*Object*", "s3:GetObject*"]
               - "s3:*MultipartUpload*"
+            Effect: Allow
+            Resource: !If [EnableEncryption, !Sub "${SynapseEncryptedExternalBucket.Arn}/*", !Sub "${SynapseExternalBucket.Arn}/*"]
+            Principal:
+              AWS: "325565585839"
+          - Sid: BucketAccess
+            Effect: Allow
+            Principal:
+              AWS: !Ref GrantAccess
+            Action:
+              - "s3:ListBucket*"
+              - "s3:GetBucketLocation"
+             Resource: !If [EnableEncryption, !GetAtt SynapseEncryptedExternalBucket.Arn, !GetAtt SynapseExternalBucket.Arn]
+          - Sid: ReadObjectAccess
+            Effect: Allow
+            Principal:
+              AWS: !Ref GrantAccess
+            Action:
+              - "s3:GetObject"
+              - "s3:GetObjectAcl"
+              - "s3:AbortMultipartUpload"
+              - "s3:ListMultipartUploadParts"
             Resource: !If [EnableEncryption, !Sub "${SynapseEncryptedExternalBucket.Arn}/*", !Sub "${SynapseExternalBucket.Arn}/*"]
+          - !If
+              - AllowWrite
+            - Sid: InternalPutObjectAccess
+              Effect: Allow
+              Principal:
+                AWS: !Ref GrantAccess
+              Action:
+                - "s3:PutObject"
+                - "s3:PutObjectAcl"
+              Resource: !If [EnableEncryption, !Sub "${SynapseEncryptedExternalBucket.Arn}/*", !Sub "${SynapseExternalBucket.Arn}/*"]
+              Condition:
+                StringLike:
+                  "aws:PrincipalArn":
+                    - "arn:aws:iam::055273631518:*"
+                    - "arn:aws:sts::055273631518:assumed-role/*"
+          - !If
+              - AllowWrite
+            - Sid: ExternalPutObjectAccess
+              Effect: Allow
+              Principal:
+                AWS: !Ref GrantAccess
+              Action:
+                - "s3:PutObject"
+                - "s3:PutObjectAcl"
+              Resource: !If [EnableEncryption, !Sub "${SynapseEncryptedExternalBucket.Arn}/*", !Sub "${SynapseExternalBucket.Arn}/*"]
+              Condition:
+                StringEquals:
+                  s3:x-amz-acl: bucket-owner-full-control
```

Comment:
these `if` conditions look wrong. The docs provide example as `!If [condition_name, value_if_true, value_if_false]` which implies that valid inputs to `If` condition are 3 strings, not lists.  Also you'll need to provide a 3rd parameter to the `If` construct, in this case i'm thinking you'll want a `AWS::NoValue`, https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/pseudo-parameter-reference.html

Reply:
> valid inputs to If condition are 3 strings, not lists

Here's an an example that uses lists:
https://stackoverflow.com/questions/56970457/how-to-use-fnif-with-array-values-in-cloud-formation-templates

There's an example on this page too:
https://github.com/aws/serverless-application-model/issues/1218

There's an example on this page that uses a nested IF in the else clause!
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/conditions-sample-templates.html

Reply:
ok, i'm convinced that it can take lists however i still think it's wrong.  The 1st example looks the most correct so i'm thinking it should be something like this..

```
  - !If
    - AllowWrite
    - - Sid: ExternalPutObjectAccess
         Effect: Allow
         Principal:
           AWS: !Ref GrantAccess
         Action:
           - "s3:PutObject"
           - "s3:PutObjectAcl"
         Resource: !If [EnableEncryption, !Sub "${SynapseEncryptedExternalBucket.Arn}/*", !Sub "${SynapseExternalBucket.Arn}/*"]
         Condition:
           StringEquals:
             s3:x-amz-acl: bucket-owner-full-control
    - - !Ref AWS::NoValue
```